runtime/include/chpl-qsbr.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2004-2018 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifndef _chpl_qsbr_h_
+#define _chpl_qsbr_h_
+
+/**
+* A Quiescent State-Based Reclamation style of Read-Copy-Update and
+* pseudo garbage-collector. We generalize and coarsen the granularity
+* of scope by keeping a single global epoch describing the state of 
+* the entire system. Updates to protected data can be 'broadcast' to the system,
+* incrementing the global epoch and moving the system's state forward.
+* Threads must periodically pass through checkpoints which indicate that they
+* are not currently accessing protected data by updating their thread-local epochs
+* to the current global epoch. Once all threads have passed some epoch, it is
+* safe for the updater to delete the older version of data. Multiple writers are
+* allowed.
+*/
+
+void chpl_qsbr_init(void);
+
+// Invoked periodically by tasks to indicate that it is no
+// longer using QSBR-Protected data. Note that this must never
+// be called while accessing the data itself. This function
+// will update TLS data. 
+void chpl_qsbr_checkpoint(void);
+
+// Broadcasts a global state change and deletes 'args' if all threads pass a
+// checkpoint. In the case that the user passes a stack-allocated array for 'args'
+// whether or not it gets deleted is optional. If a thread does not pass the checkpoint
+// the task of deletion is deferred to that thread.
+void chpl_qsbr_defer_deletion(void **data, int numData, bool deleteData);
```

Comment:
From strictly an API viewpoint, I don't know if numData is in bytes or some other unit. Also why is there a deleteData argument? Doesn't this function always (eventually) delete the data? Lastly I expect in the future we'll need it to be able to call a function to do the deletion (e.g. in case deleting the data also means deleting something it "owns" and points to).

Reply:
If it's deleteing each element in an array that is passed, that's not clear from the comment. For one the thing comment is talking about 'args' which maybe is what you mean to be the name of 'data'?

Reply:
Oh yeah, originally I wanted to allow things like `chpl_qsbr_defer_deletion(&obj, 1, false);` to allow passing a single value allocated on the stack... but then I remembered that its likely to defer deletion after the function ends and forgot to remove that argument/check, I might make it pass a single `void *` and if `numData > 1` cast it to a `void **`. `numData` is supposed to be the size of the C array.

Reply:
If you know there are common? use-cases with both deallocating a list of things and 1 thing, I think it'd be reasonable to have 2 different functions, e.g. chpl_qsbr_defer_multi_deletion and chpl_qsbr_defer_deletion.